### PR TITLE
tle6240 and external gpio chips API - preliminary

### PR DESCRIPTION
**Pull request by dron0gus**
Preliminary.
Please take a look and say if you like this style of API and driver.
Will fully test it when assemble test bench.

### CODE REVIEW

**Comment by rusefi** [comment]
Can we please-please-please split this PR into a series of smaller PR?

What is the first one to begin with? Maybe core.c?

core.c needs a unit test in my opinion. So how about we add a unit test into core.c and merge core.c with it's unit test first?

**Comment by dron0gus** [comment]
On Mon, Mar 11, 2019 at 2:03 AM rusefi <notifications@github.com> wrote:

> Can we please-please-please split this PR into a series of smaller PR?
>
> What is the first one to begin with? Maybe core.c?
>
Yes. If you ok with coding style and API.

> core.c needs a unit test in my opinion. So how about we add a unit test
> into core.c and merge core.c with it's unit test first?
>
Ok. I'll try to do some unit test for it. Never done before.

File: firmware/hw_layer/drivers/gpio/tle6240.c [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+	/* Acquire ownership of the bus. */
+	spiAcquireBus(spi);
+	/* Setup transfer parameters. */
+	spiStart(spi, &chip->cfg->spi_config);
+	/* Slave Select assertion. */
+	spiSelect(spi);
+	/* Atomic transfer operations. */
+	/* TODO: check why spiExchange transfers invalid data on STM32F7xx */
+	//spiExchange(spi, 2, txb, rxb);
+	rxb[0] = spiPolledExchange(spi, txb[0]);
```
Comment (andreika-git) [inline firmware/hw_layer/drivers/gpio/tle6240.c]: SPIv1/v2 frame is uint16_t

Comment (andreika-git) [reply firmware/hw_layer/drivers/gpio/tle6240.c]: (See SPI_CR1_DFF flag for 16-bit mode)

File: firmware/hw_layer/drivers/gpio/tle6240.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
+/*
+ * tle6240.h
+ *
+ * TLE6240GP Smart 16-Channel Low-Side Switch
+ *
+ * @date Dec 29, 2018
+ * @author Andrey Belomutskiy, (c) 2012-2018
```
Comment (rusefi) [inline firmware/hw_layer/drivers/gpio/tle6240.h]: "drivers/gpio" folder idea makes total sense. Just moving tle6240 file into it and new .mk file could be another small pull request in order to move forward

**Pull request closed without merging**